### DISCUSSION TITLE
Setting up a tenancy app

### DISCUSSION THREAD

**Discussion by peppermintbunny** [supabase/supabase#12631] [Questions]
I’m setting up a tenancy app and need help and advice on how to architecture my tenants and their users. I’m aiming to use a shared db with tenancy controlled by a column containing tenant-uuids.   Each tenant will have multiple users.  The users will be able to sign in and add records to the results table. The tenant will be able to do crud functions on the results and on other tables.   So, I have to following question:
 My tenants will register and be added to admin.users.  A trigger will add each new tenant record to public.profiles.  How should I handle the tenant’s users? So far I’ve been trying to add users to admin.users and then to public.profiles but I can’t work out how to apply policies so that there is row level security for the tenant and for each user as well as restricting what the user can do.
Advice and help will be appreciated.

**Comment by GaryAustin1** [comment]
This is one approach using the jwt:
https://github.com/supabase-community/supabase-custom-claims

Otherwise you can do joins or a function call in RLS to check your admin table for that user to see if they can do the operation.

**Reply by peppermintbunny** [reply]
Very interesting approach!.  How do I set app_metadata when I sign in a new user - I can only see how to add user_metadata?  At the moment I'm adding all my additional info to user_metadata but now I see app_metadata offers considerable advantages.

**Comment by GaryAustin1** [comment]
From server side.
https://supabase.com/docs/reference/javascript/auth-admin-updateuserbyid
You can also defaults when a user signs up in an auth trigger function.
And you can have an rpc call with service role that verifies a user has permissions to update it.
I think the repository has functions to do some of that or examples.

**Reply by GaryAustin1** [reply]
You normally should not start new answers when replying to a user...

The repository shows using functions it has for RLS, RPC and anywhere else in Postgres.
For example this shows using properties in RLS:
https://github.com/supabase-community/supabase-custom-claims#inside-an-rls-row-level-security-policy

You can certainly tweak the claims to allow a team claim and either have an admin only admin their own team (which is what I think you want.  Or have admin roles for multiple teams either with a team array or separate table.

Basically all this can work just using tables to store the "claims" also.  Using app_meta_data is just a way to "speed" up checking the claims in RLS for the current user, versus doing a table lookup in RLS for the users claims in a "profile" table.

**Comment by peppermintbunny** [comment]
I dont really want to give a user the permission or functionality to update his app_metadata.  In my multi-tenant app, the tenant creates/adds his users and the tenant deletes/destroys his users.  So, my tenant is about to use the standard "register a user" functionality to add himself to the app and create his tenancy.  This puts his record in the auth.users table.  And I could have a trigger-function to add him also to public.profiles.  I'm struggling how to deal with the users each tenant introduces.  A tenant will have an "admin app" to which only the tenant is to have access.  And the tenant will have a portal or user-app for his own users.  I'm working on the assumption that the tenant's users also have been added to auth.users and if used, public.profiles.  As developer I need to be able to direct tenants to the admin app and the tenants users to the portal or user-app.  With you help I've sorted out how to a tenant can create a user.  Now I'm struggling to get my head round differentiating between my 2 class of users (tenant and tenant's users) and hopefully take advantage of RLS.
